File: web/src/features/codex-continuation/policy.ts
export type CodexContinuationBoundary = "none" | "same_identity" | "any";

export interface CodexContinuationPolicy {
  outbound: CodexContinuationBoundary;
  inbound: CodexContinuationBoundary;
}

export const DEFAULT_CODEX_CONTINUATION: CodexContinuationPolicy = {
  outbound: "any",
  inbound: "none",
};

export function parseCodexContinuation(
  value: unknown,
): CodexContinuationPolicy {
  if (value === undefined) return { ...DEFAULT_CODEX_CONTINUATION };
  if (!value || typeof value !== "object")
    throw new Error("Invalid Codex continuation policy");
  const source = value as Record<string, unknown>;
  const parse = (boundary: unknown): CodexContinuationBoundary => {
    if (
      boundary === "none" ||
      boundary === "same_identity" ||
      boundary === "any"
    )
      return boundary;
    throw new Error("Invalid Codex continuation boundary");
  };
  return { outbound: parse(source.outbound), inbound: parse(source.inbound) };
}

File: internal/codex/continuation/policy.go
// Package continuation governs Codex conversation movement independently of
// vendor failover and of the provenance of individual upstream state values.
package continuation

import (
	"errors"
	"fmt"
)

type Boundary string

const (
	None         Boundary = "none"
	SameIdentity Boundary = "same_identity"
	Any          Boundary = "any"
)

type Policy struct {
	Outbound Boundary `json:"outbound" gorm:"default:any"`
	Inbound  Boundary `json:"inbound" gorm:"default:none"`
}

func (p Policy) Effective() Policy {
	if p.Outbound == "" {
		p.Outbound = Any
	}
	if p.Inbound == "" {
		p.Inbound = None
	}
	return p
}

func (p Policy) Validate() error {
	p = p.Effective()
	for _, boundary := range []Boundary{p.Outbound, p.Inbound} {
		if boundary != None && boundary != SameIdentity && boundary != Any {
			return fmt.Errorf("invalid Codex continuation boundary %q", boundary)
		}
	}
	return nil
}

type Route struct {
	ProviderID    string
	ProtocolScope string
	Policy        Policy
}

type Decision struct {
	Allowed          bool
	Reason           string
	SourceProviderID string
	TargetProviderID string
}

// Denied is a routing decision, not an upstream or conversation-state failure.
// Transports can reselect before visibility without charging provider health.
type Denied struct{ Decision Decision }

func (e *Denied) Error() string {
	return fmt.Sprintf("Codex continuation %s: %s -> %s", e.Decision.Reason, e.Decision.SourceProviderID, e.Decision.TargetProviderID)
}

func IsDenied(err error) bool {
	var denied *Denied
	return errors.As(err, &denied)
}

func Evaluate(source, target Route) Decision {
	d := Decision{SourceProviderID: source.ProviderID, TargetProviderID: target.ProviderID}
	sameIdentity := source.ProtocolScope != "" && source.ProtocolScope == target.ProtocolScope
	if source.ProviderID == target.ProviderID && sameIdentity {
		d.Allowed, d.Reason = true, "current_route"
		return d
	}
	if !permits(source.Policy.Effective().Outbound, sameIdentity) {
		d.Reason = "outbound_denied"
		return d
	}
	if !permits(target.Policy.Effective().Inbound, sameIdentity) {
		d.Reason = "inbound_denied"
		return d
	}
	d.Allowed, d.Reason = true, "continuation_allowed"
	return d
}

func permits(boundary Boundary, sameIdentity bool) bool {
	return boundary == Any || boundary == SameIdentity && sameIdentity
}
